File: frontend/src/api/cohort.ts
import { api } from './axios'

// ---------------- Types ----------------

export interface College {
  id: number
  name: string
  slug: string
  city: string
  state: string
  contact_name: string
  contact_email: string
  created_by?: number | null
  created_at: string
  updated_at?: string | null
}

export interface ReferralCodeOut {
  id: number
  cohort_id: number
  code: string
  max_uses: number
  used_count: number
  expires_at?: string | null
  created_at: string
}

export interface Cohort {
  id: number
  college_id: number
  course_id: number
  spoc_user_id: number
  name: string
  slug: string
  max_students: number
  starts_on?: string | null
  ends_on?: string | null
  is_active: boolean
  seat_price?: number | null
  created_at: string
  updated_at?: string | null
  referral_code?: ReferralCodeOut | null
}

export interface RosterRow {
  user_id: number
  display_name: string
  email: string
  joined_at: string
  active_today: boolean
}

export interface TaskStatsRow {
  user_id: number
  display_name: string
  completed_lessons: number
  quiz_attempts: number
  assignment_submissions: number
}

export type SessionType = 'lecture' | 'lab' | 'evaluation' | 'other'
export type AttendanceStatus = 'present' | 'absent' | 'late' | 'excused'

export interface CohortSession {
  id: number
  cohort_id: number
  scheduled_at: string
  duration_minutes: number
  topic: string
  session_type: string
  created_by?: number | null
  created_at: string
}

export interface AttendanceEntry {
  user_id: number
  status: AttendanceStatus
  notes?: string
}

// ---------------- Admin: colleges ----------------

export const collegeApi = {
  list: () => api.get<College[]>('/cohorts/admin/colleges').then(r => r.data),
  get: (id: number) => api.get<College>(`/cohorts/admin/colleges/${id}`).then(r => r.data),
  create: (body: Partial<College>) =>
    api.post<College>('/cohorts/admin/colleges', body).then(r => r.data),
  update: (id: number, body: Partial<College>) =>
    api.put<College>(`/cohorts/admin/colleges/${id}`, body).then(r => r.data),
  remove: (id: number) =>
    api.delete(`/cohorts/admin/colleges/${id}`).then(r => r.data),
}

// ---------------- Admin: cohorts ----------------

export interface CohortCreatePayload {
  college_id: number
  course_id: number
  spoc_user_id: number
  name: string
  slug?: string
  max_students?: number
  starts_on?: string | null
  ends_on?: string | null
  is_active?: boolean
  referral_max_uses?: number
  referral_expires_at?: string | null
  // Only meaningful on update (PUT) — backend CohortUpdate accepts it,
  // CohortCreate does not. null clears the override; omit to leave untouched.
  seat_price?: number | null
}

export const cohortApi = {
  list: () => api.get<Cohort[]>('/cohorts/admin/cohorts').then(r => r.data),
  get: (id: number) => api.get<Cohort>(`/cohorts/admin/cohorts/${id}`).then(r => r.data),
  create: (body: CohortCreatePayload) =>
    api.post<Cohort>('/cohorts/admin/cohorts', body).then(r => r.data),
  update: (id: number, body: Partial<CohortCreatePayload>) =>
    api.put<Cohort>(`/cohorts/admin/cohorts/${id}`, body).then(r => r.data),
  remove: (id: number) =>
    api.delete(`/cohorts/admin/cohorts/${id}`).then(r => r.data),
  // Manual cohort membership management
  listMembers: (cohortId: number) =>
    api.get<CohortMember[]>(`/cohorts/admin/cohorts/${cohortId}/members`).then(r => r.data),
  addMember: (cohortId: number, body: { email?: string; user_id?: number }) =>
    api.post<CohortMember>(`/cohorts/admin/cohorts/${cohortId}/members`, body).then(r => r.data),
  removeMember: (cohortId: number, userId: number) =>
    api.delete(`/cohorts/admin/cohorts/${cohortId}/members/${userId}`).then(r => r.data),
  // Regenerate referral code — invalidates the previous one immediately.
  regenerateCode: (cohortId: number) =>
    api.post<ReferralCodeOut>(`/cohorts/admin/cohorts/${cohortId}/referral-code/regenerate`)
      .then(r => r.data),
}

export interface CohortMember {
  membership_id: number
  user_id: number
  email: string
  display_name: string
  joined_at: string
}

// ---------------- SPOC ----------------

export const spocApi = {
  myCohorts: () => api.get<Cohort[]>('/cohorts/spoc/my-cohorts').then(r => r.data),
  roster: (cohortId: number) =>
    api.get<RosterRow[]>(`/cohorts/spoc/cohorts/${cohortId}/roster`).then(r => r.data),
  taskStats: (cohortId: number) =>
    api.get<TaskStatsRow[]>(`/cohorts/spoc/cohorts/${cohortId}/task-stats`).then(r => r.data),
  listSessions: (cohortId: number) =>
    api.get<CohortSession[]>(`/cohorts/spoc/cohorts/${cohortId}/sessions`).then(r => r.data),
  createSession: (cohortId: number, body: Partial<CohortSession>) =>
    api.post<CohortSession>(`/cohorts/spoc/cohorts/${cohortId}/sessions`, body).then(r => r.data),
  updateSession: (cohortId: number, sessionId: number, body: Partial<CohortSession>) =>
    api.put<CohortSession>(`/cohorts/spoc/cohorts/${cohortId}/sessions/${sessionId}`, body).then(r => r.data),
  deleteSession: (cohortId: number, sessionId: number) =>
    api.delete(`/cohorts/spoc/cohorts/${cohortId}/sessions/${sessionId}`).then(r => r.data),
  bulkAttendance: (sessionId: number, entries: AttendanceEntry[]) =>
    api.post(`/cohorts/spoc/sessions/${sessionId}/attendance`, { entries }).then(r => r.data),
  setEligibility: (userId: number, body: { eligible: boolean; reason?: string }) =>
    api.post(`/cohorts/spoc/students/${userId}/internship-eligible`, body).then(r => r.data),
}

File: backend/app/schemas/cohort.py
"""Pydantic v2 schemas for SS1 — cohorts module."""
from __future__ import annotations

from datetime import date, datetime
from typing import List, Optional, Literal

from pydantic import BaseModel, ConfigDict, EmailStr, Field


# ---------- College ----------

class CollegeBase(BaseModel):
    name: str
    slug: Optional[str] = None
    city: Optional[str] = ""
    state: Optional[str] = ""
    contact_name: Optional[str] = ""
    contact_email: Optional[str] = ""


class CollegeCreate(CollegeBase):
    pass


class CollegeUpdate(BaseModel):
    name: Optional[str] = None
    slug: Optional[str] = None
    city: Optional[str] = None
    state: Optional[str] = None
    contact_name: Optional[str] = None
    contact_email: Optional[str] = None


class CollegeOut(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    id: int
    name: str
    slug: str
    city: str
    state: str
    contact_name: str
    contact_email: str
    created_by: Optional[int] = None
    created_at: datetime
    updated_at: Optional[datetime] = None


# ---------- ReferralCode ----------

class ReferralCodeOut(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    id: int
    cohort_id: int
    code: str
    max_uses: int
    used_count: int
    expires_at: Optional[datetime] = None
    created_at: datetime


# ---------- Cohort ----------

class CohortBase(BaseModel):
    college_id: int
    course_id: int
    spoc_user_id: int
    name: str
    slug: Optional[str] = None
    max_students: int = 0
    starts_on: Optional[date] = None
    ends_on: Optional[date] = None
    is_active: bool = True


class CohortCreate(CohortBase):
    referral_max_uses: int = Field(default=100, ge=1)
    referral_expires_at: Optional[datetime] = None


class CohortUpdate(BaseModel):
    name: Optional[str] = None
    slug: Optional[str] = None
    college_id: Optional[int] = None
    course_id: Optional[int] = None
    spoc_user_id: Optional[int] = None
    max_students: Optional[int] = None
    starts_on: Optional[date] = None
    ends_on: Optional[date] = None
    is_active: Optional[bool] = None
    # Per-cohort seat price override (e.g. college-negotiated pricing).
    # Explicit null clears the override back to the course's normal price;
    # omitting the field leaves it untouched (handler uses exclude_unset).
    seat_price: Optional[float] = Field(default=None, gt=0)


class CohortOut(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    id: int
    # college_id / course_id are nullable in the DB model (internship-spawned
    # cohorts have no college or course). A previous non-Optional schema
    # caused /admin/cohorts to 500 on Pydantic validation when any such
    # cohort existed.
    college_id: Optional[int] = None
    course_id: Optional[int] = None
    spoc_user_id: int
    name: str
    slug: str
    max_students: int
    starts_on: Optional[date] = None
    ends_on: Optional[date] = None
    is_active: bool
    seat_price: Optional[float] = None
    created_at: datetime
    updated_at: Optional[datetime] = None
    referral_code: Optional[ReferralCodeOut] = None


# ---------- Roster / task stats ----------

class RosterRow(BaseModel):
    user_id: int
    display_name: str
    email: str
    joined_at: datetime
    active_today: bool  # activity-derived from LessonProgress / QuizAttempt


class TaskStatsRow(BaseModel):
    user_id: int
    display_name: str
    completed_lessons: int
    quiz_attempts: int
    assignment_submissions: int


# ---------- Sessions / attendance ----------

SessionType = Literal["lecture", "lab", "evaluation", "other"]
AttendanceStatus = Literal["present", "absent", "late", "excused"]


class SessionBase(BaseModel):
    scheduled_at: datetime
    duration_minutes: int = 60
    topic: str = ""
    session_type: SessionType = "lecture"


class SessionCreate(SessionBase):
    pass


class SessionUpdate(BaseModel):
    scheduled_at: Optional[datetime] = None
    duration_minutes: Optional[int] = None
    topic: Optional[str] = None
    session_type: Optional[SessionType] = None


class SessionOut(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    id: int
    cohort_id: int
    scheduled_at: datetime
    duration_minutes: int
    topic: str
    session_type: str
    created_by: Optional[int] = None
    created_at: datetime


class AttendanceEntry(BaseModel):
    user_id: int
    status: AttendanceStatus
    notes: Optional[str] = ""


class BulkAttendanceRequest(BaseModel):
    entries: List[AttendanceEntry]


class AttendanceOut(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    id: int
    session_id: int
    user_id: int
    status: str
    notes: str
    marked_by: Optional[int] = None
    marked_at: Optional[datetime] = None


# ---------- Eligibility toggle (SS2 handoff) ----------

class EligibilityToggleRequest(BaseModel):
    eligible: bool
    reason: Optional[str] = ""


class EligibilityToggleOut(BaseModel):
    user_id: int
    eligible: bool
    source: str
    cohort_id: Optional[int] = None
    decided_by: Optional[int] = None
    reason: str = ""
